### PR TITLE
WIP: Portfolio 1.5

### PR DESCRIPTION
**Pull request by jsom**

### CODE REVIEW

File: ExampleAppUITests/TradeItSDK2UITestsPortfolioFlow.swift [hunk 1]
```diff
@@ -123,63 +123,4 @@ class TradeItSDK2UITestsPortfolioFlow: XCTestCase {
         XCTAssertFalse(app.tables.staticTexts["Margin Acc**cct4"].exists)
         XCTAssertFalse(app.tables.staticTexts["OPTIONS SU**cct5"].exists)
     }
-
```
Comment (jsom) [inline ExampleAppUITests/TradeItSDK2UITestsPortfolioFlow.swift]: These have been commented for a while. Seeing as I'm stripping out other FX code for now it seems the right time to remove this too.

File: TradeItIosTicketSDK2/NumberFormatter.swift [hunk 2]
```diff
@@ -1,29 +1,43 @@
...
-        return formatter.string(from: number) ?? TradeItPresenter.MISSING_DATA_PLACEHOLDER
+        currencyFormatter.numberStyle = .currency
+        currencyFormatter.currencyCode = currencyCode
+        currencyFormatter.currencySymbol = overrideCurrencySymbol(forCurrencyCode: currencyCode)
+        if variance {
+            currencyFormatter.positivePrefix = "+" + currencyFormatter.currencySymbol
+            currencyFormatter.negativePrefix = "-" + currencyFormatter.currencySymbol
+            // TODO: See if we can get this working
+//            currencyFormatter.textAttributesForPositiveValues = [NSForegroundColorAttributeName: UIColor.tradeItBuyGreenColor]
+//            currencyFormatter.textAttributesForNegativeValues = [NSForegroundColorAttributeName: UIColor.tradeItDeepRoseColor]
```
Comment (jsom) [inline TradeItIosTicketSDK2/NumberFormatter.swift]: This doesn't seem to work. It should colour the text based on pos/neg.

File: TradeItIosTicketSDK2/NumberFormatter.swift [hunk 3]
```diff
@@ -1,29 +1,43 @@
 class NumberFormatter: NSObject {
-    private static let formatter = Foundation.NumberFormatter()
+    private static let currencyFormatter = Foundation.NumberFormatter()
+    private static let quantityFormatter = Foundation.NumberFormatter()
+    private static let percentageFormatter = Foundation.NumberFormatter()
```
Comment (jsom) [inline TradeItIosTicketSDK2/NumberFormatter.swift]: These are expensive to create so need to reuse them.

File: TradeItIosTicketSDK2/TradeItAccountSummaryView.swift [hunk 4]
```diff
@@ -2,19 +2,10 @@ import UIKit
...
-            self.accountEquitySummaryView.isHidden = false
-            self.accountEquitySummaryView.populate(withAccount: selectedAccount)
-        }
-        else if selectedAccount.fxBalance != nil {
-            self.accountEquitySummaryView.isHidden = true
-            self.accountFxSummaryView.isHidden = false
-            self.accountFxSummaryView.populate(withAccount: selectedAccount)
-        }
+        self.accountEquitySummaryView.isHidden = false
+        self.accountEquitySummaryView.populate(withAccount: selectedAccount)
```
Comment (jsom) [inline TradeItIosTicketSDK2/TradeItAccountSummaryView.swift]: Removing FX balances altogether for now.

File: TradeItIosTicketSDK2/TradeItPortfolioAccountDetailsTableViewManager.swift [hunk 5]
```diff
@@ -136,14 +174,9 @@ class TradeItPortfolioPositionsTableViewManager: NSObject, UITableViewDelegate,
             equityCell.showPositionDetails(selected)
             cell = equityCell
         } else if position.fxPosition != nil {
-            let fxCell = tableView.dequeueReusableCell(withIdentifier: "PORTFOLIO_FX_POSITIONS_CELL_ID") as! TradeItPortfolioFxPositionsTableViewCell
-            fxCell.populate(withPosition: position)
-            fxCell.showPositionDetails(selected)
-            cell = fxCell
+            return UITableViewCell()
```
Comment (jsom) [inline TradeItIosTicketSDK2/TradeItPortfolioAccountDetailsTableViewManager.swift]: Don't handle FX for now.

File: TradeItIosTicketSDK2/TradeItPortfolioAccountDetailsViewController.swift [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+        order.action = orderAction ?? TradeItOrderActionPresenter.DEFAULT
+        return order
+    }
+
+    // MARK: IBActions
+
+    // TODO: Replace with Accounts button up top
+    //    @IBAction func editAccountsButtonTapped(_ sender: UIButton) {
+    //        TradeItSDK.launcher.launchAccountManagement(fromViewController: self)
+    //    }
```
Comment (jsom) [inline TradeItIosTicketSDK2/TradeItPortfolioAccountDetailsViewController.swift]: This still needs to be done.

File: TradeItIosTicketSDK2/TradeItPortfolioAccountDetailsViewController.swift [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+
+    @IBAction func closeButtonTapped(_ sender: UIBarButtonItem) {
+        self.parent?.dismiss(animated: true, completion: nil)
+    }
+
+    // TODO: Move to detail view
+    //    @IBAction func tradeButtonWasTapped(_ sender: AnyObject) {
+    //        let order = provideOrder(forPortFolioPosition: nil, account: self.selectedAccount, orderAction: nil)
+    //        self.tradingUIFlow.presentTradingFlow(fromViewController: self, withOrder: order)
+    //    }
```
Comment (jsom) [inline TradeItIosTicketSDK2/TradeItPortfolioAccountDetailsViewController.swift]: Do we still want this Trade button?

File: TradeItIosTicketSDK2/TradeItPortfolioAccountTableViewCell.swift [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+    func populate(withAccount account: TradeItLinkedBrokerAccount) {
+        let presenter = TradeItPortfolioBalanceEquityPresenter(account)
+        self.accountNameLabel.text = account.getFormattedAccountName()
+        self.totalValueLabel.text = presenter.getFormattedTotalValue()
+        self.returnLabel.text = presenter.getFormattedDayReturnWithPercentage()
+        self.returnLabel.textColor = TradeItPresenter.stockChangeColor(account.balance?.dayAbsoluteReturn?.doubleValue)
+        self.setFailedState(account: account)
+    }
+
+    private func setFailedState(account: TradeItLinkedBrokerAccount) {
```
Comment (jsom) [inline TradeItIosTicketSDK2/TradeItPortfolioAccountTableViewCell.swift]: Probably should rename this... `handleFailedState`?

File: ExampleApp/AppDelegate.swift [hunk 9]
```diff
@@ -4,7 +4,8 @@ import TradeItIosTicketSDK2
 @UIApplicationMain
 class AppDelegate: UIResponder, UIApplicationDelegate {
     static let API_KEY = "tradeit-test-api-key" //"tradeit-fx-test-api-key"
-    static let ENVIRONMENT = TradeItEmsTestEnv
+//    static let ENVIRONMENT = TradeItEmsTestEnv
+    static let ENVIRONMENT = TradeItEmsProductionEnv
```
Comment (jsom) [inline ExampleApp/AppDelegate.swift]: Will revert this change

File: TradeItIosTicketSDK2/TradeItPortfolioAccountDetailsTableViewManager.swift [hunk 10]
```diff
@@ -50,51 +61,72 @@ class TradeItPortfolioPositionsTableViewManager: NSObject, UITableViewDelegate,
...
-                                    forPortfolioPosition: position,
-                                    selected: self.selectedPositionIndex == indexPath.row)
-        return cell
+        if indexPath.section == SECTIONS.accountDetails.rawValue {
+            if let account = self.account {
+                let cell = tableView.dequeueReusableCell(withIdentifier: "TRADE_IT_PORTFOLIO_ACCOUNT_DETAILS") as! TradeItPortfolioAccountDetailsTableViewCell
+                cell.populate(withAccount: account)
+                return cell
+            } else {
+                return UITableViewCell()
```
Comment (jsom) [inline TradeItIosTicketSDK2/TradeItPortfolioAccountDetailsTableViewManager.swift]: ???

File: TradeItIosTicketSDK2/TradeItPortfolioAccountDetailsViewController.swift [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+            linkedBrokerAccount.getPositions(
+                onSuccess: { positions in
+                    self.activityView?.hide(animated: true)
+                    self.tableViewManager.updatePositions(withPositions: positions)
+                }, onFailure: { errorResult in
+                    self.activityView?.hide(animated: true)
+                    // TODO: Figure out error handling
+                    print(errorResult)
+                }
+            )
```
Comment (jsom) [inline TradeItIosTicketSDK2/TradeItPortfolioAccountDetailsViewController.swift]: I didn't make this promise based because they are both distinct requests - although it would be nice to have a single error handler.

File: TradeItIosTicketSDK2/TradeItPortfolioAccountsTableViewManager.swift [hunk 12]
```diff
@@ -25,74 +20,132 @@ class TradeItPortfolioAccountsTableViewManager: NSObject, UITableViewDelegate, U
...
     }
 
-    func tableView(_ tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
-        return 30.0
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        if (section == 0) {
+            return 1
+        } else {
+            guard let linkedBrokerSectionPresenter = self.linkedBrokerSectionPresenters[safe: section - 1] else { return 0 }
+            return linkedBrokerSectionPresenter.numberOfRows()
```
Comment (jsom) [inline TradeItIosTicketSDK2/TradeItPortfolioAccountsTableViewManager.swift]: This is incorrect if the broker has an error. In that case it should have +1 row.

File: TradeItIosTicketSDK2/TradeItPortfolioAccountsTableViewManager.swift [hunk 13]
```diff
@@ -25,74 +20,132 @@ class TradeItPortfolioAccountsTableViewManager: NSObject, UITableViewDelegate, U
...
+
+    func cellFor(tableView: UITableView, andRow row: Int) -> UITableViewCell {
+        if row == 0 && hasError() {
+            guard let error = linkedBroker.error else { return UITableViewCell() } // TODO: Wut?
+            let cell = tableView.dequeueReusableCell(withIdentifier: "TRADE_IT_PORTFOLIO_LINKED_BROKER_ERROR") as! TradeItPortfolioLinkedBrokerErrorTableViewCell
+            cell.populate(withError: error)
+            return cell
         }
 
+        guard let account = accountFor(row: row) else { return UITableViewCell() }
```
Comment (jsom) [inline TradeItIosTicketSDK2/TradeItPortfolioAccountsTableViewManager.swift]: This needs to be based on whether or not there is an error row above.

File: TradeItIosTicketSDK2/TradeItPortfolioAccountsTableViewManager.swift [hunk 14]
```diff
@@ -25,74 +20,132 @@ class TradeItPortfolioAccountsTableViewManager: NSObject, UITableViewDelegate, U
...
         return cell
     }
+
+    func accountFor(row: Int) -> TradeItLinkedBrokerAccount? {
+        return self.linkedBroker.accounts[safe: row - errorOffset()]
+    }
+
+    func hasError() -> Bool {
+        guard let error = linkedBroker.error else { return false }
+        return !error.requiresAuthentication()
```
Comment (jsom) [inline TradeItIosTicketSDK2/TradeItPortfolioAccountsTableViewManager.swift]: Need to review logic here...

File: TradeItIosTicketSDK2/TradeItPortfolioEquityPositionPresenter.swift [hunk 15]
```diff
@@ -34,7 +34,7 @@ class TradeItPortfolioEquityPositionPresenter: TradeItPortfolioPositionPresenter
     func getFormattedTotalReturn() -> String {
         guard let totalGainLossDollars = self.position?.totalGainLossDollar
             else { return TradeItPresenter.MISSING_DATA_PLACEHOLDER }
-        return TradeItPresenter.indicator(totalGainLossDollars.doubleValue) + " " + "\(formatCurrency(totalGainLossDollars))(\(returnPercent()))";
+        return "\(formatCurrency(totalGainLossDollars)) (\(returnPercent()))";
```
Comment (jsom) [inline TradeItIosTicketSDK2/TradeItPortfolioEquityPositionPresenter.swift]: Should use the same format as `TradeItPortfolioBalanceEquityPresenter`

File: TradeItIosTicketSDK2/TradeItPortfolioAccountsTableViewManager.swift (hunk 13 above)
Comment (jsom) [inline TradeItIosTicketSDK2/TradeItPortfolioAccountsTableViewManager.swift]: @mitochondrion REMEMBER THIS TOMORROW

File: TradeItIosTicketSDK2/NumberFormatter.swift [hunk 16]
```diff
@@ -1,29 +1,40 @@
...
     
-    static func formatCurrency(_ number: NSNumber, maximumFractionDigits: Int = 2, currencyCode: String?) -> String {
+    static func formatCurrency(_ number: NSNumber, maximumFractionDigits: Int = 2, variance: Bool = false, currencyCode: String?) -> String {
```
Comment (mitochondrion) [inline TradeItIosTicketSDK2/NumberFormatter.swift]: - `variance` -> `displayVariance`
- Think about splitting formatting and presentation into the appropriate formatter/presenter.

File: TradeItIosTicketSDK2/TradeItPortfolioAccountDetailsTableViewManager.swift (hunk 10 above)
Comment (mitochondrion) [inline TradeItIosTicketSDK2/TradeItPortfolioAccountDetailsTableViewManager.swift]: Change this to `guard let account = self.account else { return UITableViewCell() }` or whatever...

File: TradeItIosTicketSDK2/TradeItPortfolioAccountDetailsViewController.swift (hunk 6 above)
Comment (mitochondrion) [inline TradeItIosTicketSDK2/TradeItPortfolioAccountDetailsViewController.swift]: 👌

File: TradeItIosTicketSDK2/TradeItPortfolioAccountDetailsViewController.swift (hunk 7 above)
Comment (mitochondrion) [inline TradeItIosTicketSDK2/TradeItPortfolioAccountDetailsViewController.swift]: 👌

File: TradeItIosTicketSDK2/TradeItPortfolioAccountsTableViewManager.swift (hunk 12 above)
Comment (mitochondrion) [inline TradeItIosTicketSDK2/TradeItPortfolioAccountsTableViewManager.swift]: 👌

File: TradeItIosTicketSDK2/TradeItPortfolioAccountsTableViewManager.swift (hunk 13 above)
Comment (mitochondrion) [inline TradeItIosTicketSDK2/TradeItPortfolioAccountsTableViewManager.swift]: 👌
